### ISSUE TITLE
Using Query causes object assign errors

### ISSUE THREAD

**Issue by Reklino** [apollographql/react-apollo#2609]
**Intended outcome:**
Using `Query` in React Native Expo project should return data as expected.

I just upgraded to [Expo 31](https://medium.com/@wcandillon/congratulations-expo-and-react-native-and-incredible-tools-c83dfb143599), and I suspect this may have something to do with babel 7 and the spread operator, but I'm not sure.

``` javascript
<Query query={gql `
  {
    drafts {
      id
      title
    }
  }`}>
  {({ loading, error, data }) => {
    console.log(error)
    if (loading) return null;
    if (error) return <Text>{`Error!: ${error}`}</Text>;
    
    return (
      <Text>{ data.drafts.length }</Text>
    );
  }}
</Query>
```

**Actual outcome:**
Using `Query` returns an error:
```
Network error: In this environment the sources for assign MUST be an object. This error is a performance optimization and not spec compliant.

Stack trace:
  node_modules/apollo-client/bundle.umd.js:85:36 in ApolloError
  node_modules/apollo-client/bundle.umd.js:194:43 in currentResult
  node_modules/react-apollo/react-apollo.browser.umd.js:350:40 in getQueryResult
  node_modules/react-apollo/react-apollo.browser.umd.js:470:67 in finish
  node_modules/react-native/Libraries/Renderer/oss/ReactNativeRenderer-dev.js:10563:21 in finishClassComponent
  node_modules/react-native/Libraries/Renderer/oss/ReactNativeRenderer-dev.js:14091:21 in performUnitOfWork
  node_modules/react-native/Libraries/Renderer/oss/ReactNativeRenderer-dev.js:14129:41 in workLoop
  node_modules/react-native/Libraries/Renderer/oss/ReactNativeRenderer-dev.js:14226:15 in renderRoot
  node_modules/react-native/Libraries/Renderer/oss/ReactNativeRenderer-dev.js:15193:17 in performWorkOnRoot
  node_modules/react-native/Libraries/Renderer/oss/ReactNativeRenderer-dev.js:15090:24 in performWork
  node_modules/react-native/Libraries/Renderer/oss/ReactNativeRenderer-dev.js:15047:14 in performSyncWork
  node_modules/react-native/Libraries/Renderer/oss/ReactNativeRenderer-dev.js:14925:19 in requestWork
  node_modules/react-native/Libraries/Renderer/oss/ReactNativeRenderer-dev.js:14711:16 in scheduleWork
  node_modules/react-native/Libraries/Renderer/oss/ReactNativeRenderer-dev.js:7737:17 in enqueueForceUpdate
  node_modules/react/cjs/react.development.js:382:34 in forceUpdate
  node_modules/react-apollo/react-apollo.browser.umd.js:341:20 in updateCurrentData
  node_modules/react-apollo/react-apollo.browser.umd.js:329:24 in error
  node_modules/zen-observable/lib/Observable.js:134:22 in notifySubscription
  node_modules/zen-observable/lib/Observable.js:165:21 in onNotify
  node_modules/zen-observable/lib/Observable.js:224:15 in error
  node_modules/apollo-client/bundle.umd.js:442:44 in error
  node_modules/apollo-client/bundle.umd.js:1046:47 in <unknown>
  node_modules/apollo-client/bundle.umd.js:1411:29 in <unknown>
  node_modules/apollo-client/bundle.umd.js:1410:29 in <unknown>
  node_modules/apollo-client/bundle.umd.js:1405:33 in broadcastQueries
  node_modules/apollo-client/bundle.umd.js:988:28 in <unknown>
  node_modules/promise/setimmediate/core.js:37:14 in tryCallOne
  node_modules/promise/setimmediate/core.js:123:25 in <unknown>
  node_modules/react-native/Libraries/Core/Timers/JSTimers.js:152:14 in _callTimer
  node_modules/react-native/Libraries/Core/Timers/JSTimers.js:200:17 in _callImmediatesPass
  node_modules/react-native/Libraries/Core/Timers/JSTimers.js:464:30 in callImmediates
  node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:320:6 in __callImmediates
  node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:135:6 in <unknown>
  node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:297:10 in __guard
  node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:134:17 in flushedQueue
```

**How to reproduce the issue:**

package.json
``` json
{
  "name": "my-new-project",
  "main": "node_modules/expo/AppEntry.js",
  "private": true,
  "scripts": {
    "start": "expo start",
    "android": "expo start --android",
    "ios": "expo start --ios",
    "eject": "expo eject",
    "test": "node ./node_modules/jest/bin/jest.js --watchAll"
  },
  "jest": {
    "preset": "jest-expo"
  },
  "dependencies": {
    "@expo/samples": "2.1.1",
    "apollo-cache-inmemory": "1.3.9",
    "apollo-client": "2.4.5",
    "apollo-link-context": "1.0.9",
    "apollo-link-http": "1.5.5",
    "expo": "^31.0.4",
    "graphql": "14.0.2",
    "graphql-tag": "2.10.0",
    "react": "16.5.0",
    "react-apollo": "2.3.2-beta.0",
    "react-dom": "^16.6.3",
    "react-native": "https://github.com/expo/react-native/archive/sdk-31.0.1.tar.gz",
    "react-navigation": "3.0.0"
  },
  "devDependencies": {
    "babel-plugin-add-react-displayname": "0.0.5",
    "babel-plugin-graphql-tag": "1.6.0",
    "babel-plugin-transform-react-jsx-source": "6.22.0",
    "babel-plugin-transform-runtime": "6.23.0",
    "jest-expo": "^31.0.0"
  }
}
```

babel.config.js
``` javascript
module.exports = function (api) {
    console.log(api)
    api.cache(true);
    return {
        presets: ['babel-preset-expo']
};
```

**Version**

- apollo-client@2.4.5
- react-apollo@2.3.2-beta.0

**Comment by Reklino** [comment]
So, this actually turned out to be an issue with my initial `ApolloClient` config. I wasn't supplying a valid `link`. Not sure why the error didn't point me in that direction, though 🤔. Hope that helps @maxschmeling!

**Comment by maxschmeling** [comment]
@Reklino Hah, I forgot I commented here. It actually turned out to be a bug in my code as well. It took a lot of debugging to track it down though because the error didn't point me to the right place at all. 

Eventually found the `Object.assign` implementation that was being called, put a breakpoint inside it, and watched until I found an invalid value (a string) being passed in and then checked the call stack to see where it was coming from. Then it was immediately obvious what I had done wrong. I had a method with a condition in it where one side of the condition returned the wrong value (a string instead of the full object).
